src/darwin/Framework/CHIP/MTRDevice_Internal.h

```diff
@@ -28,13 +28,18 @@ NS_ASSUME_NONNULL_BEGIN
 
 typedef void (^MTRDevicePerformAsyncBlock)(MTRBaseDevice * baseDevice);
 
+// Whether to store attributes by cluster instead of as individual entries for each attribute
+#define MTRDEVICE_ATTRIBUTE_CACHE_STORE_ATTRIBUTES_BY_CLUSTER 1
+
 /**
  * Information about a cluster, currently is just data version
  */
 MTR_TESTABLE
-@interface MTRDeviceClusterData : NSObject <NSSecureCoding>
+@interface MTRDeviceClusterData : NSObject <NSSecureCoding, NSCopying>
 @property (nonatomic) NSNumber * dataVersion;
-// TODO: add cluster attributes in this object, and remove direct attribute storage
+#if MTRDEVICE_ATTRIBUTE_CACHE_STORE_ATTRIBUTES_BY_CLUSTER
+@property (nonatomic) NSDictionary<NSNumber *, NSDictionary *> * attributes; // attributeID => data-value dictionary
```

Comment:
But we touch this property unconditionally in the impl, no?  Does this compile with MTRDEVICE_ATTRIBUTE_CACHE_STORE_ATTRIBUTES_BY_CLUSTER not set?

Reply:
You're right - I've removed it locally because I found the same when I tried to make sure things still work with the define off.

I also just decided to remove this `#if` instance because it'd be too many ugly `#if`s in the impl file if this is defined out.